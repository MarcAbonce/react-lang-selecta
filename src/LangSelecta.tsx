import React from 'react';

import _localeRegions from './data/locale_regions.json';

type LocaleRegions = {
  [locale: string]: {
    name: string;
    regions: string[];
  }
};

const localeRegions: LocaleRegions = _localeRegions;

type LangSelectaProps = {
  langs: string[];
};

function getFlagEmoji(regionCode: string) {
  const codePoints = regionCode.split('').map((char) => 127397 + char.charCodeAt(0));
  return String.fromCodePoint(...codePoints);
}

function pickRandomFlagEmoji(lang: string) {
  if (lang in localeRegions) {
    const regions = localeRegions[lang].regions;
    const index = Math.floor(Math.random() * regions.length);
    return getFlagEmoji(regions[index]);
  } else {
    // return United Nations flag as a fallback
    return "ðŸ‡ºðŸ‡³";
  }
}

function LangSelecta({ langs } : LangSelectaProps) {
  const langsData = langs.map(lang => ({
    code: lang,
    name: lang in localeRegions ? localeRegions[lang].name : lang,
    flag: pickRandomFlagEmoji(lang),
  }));

  return (
    <select>
      {langsData.map(lang => (
        <option value={lang.code}>{lang.flag} {lang.name}</option>
      ))}
    </select>
  );
}

export default LangSelecta;
