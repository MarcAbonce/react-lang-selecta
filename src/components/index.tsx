export { LangSelecta } from './LangSelecta'
export type { LangSelectaProps } from './LangSelecta'
