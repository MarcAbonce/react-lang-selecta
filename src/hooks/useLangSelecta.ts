/*
 * ðŸ„¯ 2023 Marc Abonce Seguin
 *
 * SPDX-License-Identifier: LiLiQ-R-1.1 âšœ
 * License-Filename: LICENCE.txt
 */

import { useMemo } from "react";
import leftPad from "left-pad";

import { localeRegions } from "../data";
import { getRandomListIndex } from "../utils/getRandomListIndex";
import { shuffleList } from "../utils/shuffleList";
import type { ExtraLangData } from "../types";
import { pickRandomFlagEmoji } from "../utils/pickRandomFlagEmoji";

export interface LangSelectaHookProps {
  langs: string[];
  extraLangData?: ExtraLangData;
}

export interface LangSelectaHookResult {
  langsData: {
    code: string;
    name: string;
    flag: string;
    optionName: string;
  }[];
}

export const useLangSelecta = ({
  langs,
  extraLangData,
}: LangSelectaHookProps): LangSelectaHookResult => {
  const langsData = useMemo(
    () =>
      shuffleList(langs).map((lang) => {
        let langCode, name, flag;
        // extraLangData can override localeRegions
        if (extraLangData !== undefined && lang in extraLangData) {
          const langFlags = extraLangData[lang].flags;
          flag = langFlags[getRandomListIndex(langFlags.length)];
          langCode = lang;
          name = extraLangData[lang].name;
        } else {
          langCode = lang.split("-")[0];
          flag = pickRandomFlagEmoji(langCode);
          name =
            langCode in localeRegions ? localeRegions[langCode].name : lang;
        }

        const optionName = flag + leftPad(name, name.length + 1);

        return {
          code: langCode,
          name,
          flag,
          optionName,
        };
      }),
    [langs, extraLangData],
  );

  return { langsData };
};
