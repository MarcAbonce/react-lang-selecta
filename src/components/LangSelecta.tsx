/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { forwardRef, type ComponentPropsWithoutRef } from 'react'

import { useLangSelecta } from '../hooks/useLangSelecta'
import type { ExtraLangData } from '../types'

export type LangSelectaProps = {
  langs: string[]
  extraLangData?: ExtraLangData
} & ComponentPropsWithoutRef<'select'>

const LangSelecta = forwardRef<HTMLSelectElement, LangSelectaProps>(
  ({ langs, extraLangData, ...rest }, ref): JSX.Element => {
    const { langsData } = useLangSelecta({ langs, extraLangData })

    return (
      <select ref={ref} {...rest}>
        {langsData.map(lang => (
          <option key={lang.code} value={lang.code}>{lang.optionName}</option>
        ))}
      </select>
    )
  }
)

LangSelecta.displayName = 'LangSelecta'

export { LangSelecta }
