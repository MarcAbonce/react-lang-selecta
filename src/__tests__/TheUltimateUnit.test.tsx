/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * 
 * Especially this Masterpiece. ðŸŽ“ðŸ“œ
 */

import { render } from '@testing-library/react'

import { LangSelecta } from '../components/LangSelecta'
import { useLangSelecta } from '../hooks/useLangSelecta'

// for some mysterious reason mockImplementation didn't work
// so I'll just mock this over here like a savage
jest.mock('../components/LangSelecta', () => ({
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  LangSelecta: (props: any) => {
    mockUseLangSelecta(props)
    return <div />
  }
}))
jest.mock('../hooks/useLangSelecta')

const mockUseLangSelecta = jest.mocked(useLangSelecta)

describe('The ultimate unit test', () => {
  describe('when everything is mocked', () => {
    beforeEach(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      mockUseLangSelecta.mockReturnValue(true as any)
    })

    it('does not change', () => {
      expect(render(<LangSelecta langs={['taq']}/>).container).toMatchSnapshot()
    })

    it('mock component calls mock hook', () => {
      render(<LangSelecta langs={[]} />)
      expect(mockUseLangSelecta).toHaveBeenCalledWith({ langs: [] })
    })

  })
})
