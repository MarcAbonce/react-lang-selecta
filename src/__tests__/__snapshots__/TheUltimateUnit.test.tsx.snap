// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ultimate unit test when everything is mocked does not change 1`] = `
<div>
  <div />
</div>
`;
